flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+			referenceStates, snapshot1, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot1.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(5, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(1, stateMap.getHighestFinishedSnapshotVersion());
+
+		// complete snapshot 3, and v0 will be removed
+		referenceStates.get(1L).put(1, "2");
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot3, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot3.release();
+		referenceValues.remove(referenceValues.size() - 1);
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(4, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(1, stateMap.getHighestFinishedSnapshotVersion());
+
+		// complete snapshot 2, and no value will be removed
+		referenceStates.get(1L).put(1, "1");
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot2, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot2.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(4, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(3, stateMap.getHighestFinishedSnapshotVersion());
+
+		// add node to pruning set to prevent snapshot4 to prune
+		stateMap.getPruningValueNodes().add(node);
+		// complete snapshot 4, and no value will be removed
+		referenceStates.get(1L).put(1, "3");
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot4, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot4.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(4, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(4, stateMap.getHighestFinishedSnapshotVersion());
+
+		stateMap.getPruningValueNodes().remove(node);
+
+		// complete snapshot 5, v1 and v2 will be removed
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot5, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot5.release();
+		referenceValues.remove(referenceValues.size() - 1);
+		referenceValues.remove(referenceValues.size() - 1);
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(2, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(5, stateMap.getHighestFinishedSnapshotVersion());
+
+		// complete snapshot 6, no value will be removed
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot6, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot6.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(2, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(6, stateMap.getHighestFinishedSnapshotVersion());
+
+		assertEquals("3", stateMap.removeAndGetOld(1, 1L));
+		assertEquals(0 , stateMap.size());
+		assertEquals(0, spaceAllocator.getTotalSpaceNumber());
+
+		stateMap.close();
```

Comment:
I think this test mixes different tests and could be simplified by splitting it. For example, the test that a completing snapshot does not prune if the node is already contained in the `getPruningValueNodes` could be factored out (apart from that I don't like this kind of test because it tests some internal details). If we want to test such a behaviour, then this should happen through the public APIs.

Reply:
Putting nodes into `getPruningValueNodes` is for simulating that two in-parallel snapshots won't prune a same node twice, it's like snapshot5 has put the node into pruning nodes list first and will take responsibility to remove it. I think the test case here is constructing a snapshot-update-crossing event serial to cover all possible "snapshot prune values" condition, it's relatively short with clear logic so should be easy to read and analyze when failure occurs.